### ISSUE TITLE
super().__init__(....) in mixins fails with `Too many arguments for "__init__" of "object" `

### ISSUE THREAD

**Issue by kurtgn** [python/mypy#5887]
When I create a mixin class that extends the logic of `__init__`,  the regular thing to do is:

```
class ExtraValuemixin:

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # some extra initialization
        
```

However this looks wrong to mypy, as it says:
```
Too many arguments for "__init__" of "object"
```

I get it, there's no `*args` or `**kwargs` in the `object`'s constructor signature; but this is a mixin, and it relies on its childen's constructors. Ho do I make mypy understand this?

Full example:

```
class ExtraValuemixin:

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # some extra initialization
        self._extra_value = 1

    def retrieve_extra_value(self):
        return self._extra_value


class ParentObj:

    def __init__(self, value):
        self.value = value


class ChildObj(ExtraValuemixin, ParentObj):
    pass


obj = ChildObj(value=5)

print(obj.retrieve_extra_value())
```

**Comment by mathfac** [comment]
Faced with the same issue.

**Comment by JukkaL** [comment]
Mypy shouldn't complain about this idiom, but I'm not sure what exact rules should be. Maybe calls to `super().__init__` should be special cased.

**Comment by gvanrossum** [comment]
There are lots of problems with mixins. Maybe we should have a way to flag them?

**Comment by ilevkivskyi** [comment]
@gvanrossum 
> There are lots of problems with mixins. Maybe we should have a way to flag them?

There is a proposal https://github.com/python/typing/issues/246 (partially based on https://github.com/python/typing/issues/241)

**Comment by gvanrossum** [comment]
Hmm... Maybe it’s time to work on both of these.

**Comment by ilevkivskyi** [comment]
> Hmm... Maybe it’s time to work on both of these.

It would be great to do it this year, I am not sure however we will have time to do this soon. Since it may be a relatively time-consuming change (taking into account necessary design discussions).

**Comment by NeilGirdhar** [comment]
This doesn't look like it's just a problem with mixins.  This looks like a special case of: https://github.com/python/mypy/issues/4001

Other duplicates:
- https://github.com/python/mypy/issues/4997
- https://github.com/python/mypy/issues/4335

This approach of forwarding args and kwargs seems pretty common and it would be nice if it didn't trigger any errors.

**Comment by ZdenekM** [comment]
It would be great if I could somehow specify, with what classes the mixin is going to be mixed up. Then, mypy might be able to judge calls to parent methods (including `__init__`) properly.

**Comment by kaushikcfd** [comment]
On `mypy==0.920`, the provided MFE doesn't fail for me.

**Comment by LefterisJP** [comment]
Going from mypy==0.910 to mypy=0.930 this seems to be fixed. Got bunch of type:ignore commments not needed. So I would say this issue is fixed :smile:

**Comment by erictraut** [comment]
Yes, I can confirm this appears to be fixed in the latest release.

**Comment by marcinocto** [comment]
I'm still getting `error: Too many arguments for "__init__" of "object"` in a scenario where one of the arguments is verbose with mypy==0.931
```py
class ExtraValuemixin:
    def __init__(self, value, *args, **kwargs):
        super().__init__(value, *args, **kwargs)

    def retrieve_extra_value(self):
        return self._extra_value


class ParentObj:
    def __init__(self, value):
        self.value = value


class ChildObj(ExtraValuemixin, ParentObj):
    pass


obj = ChildObj(value=5)

print(obj.retrieve_extra_value())
```
My usecase is adding mixins to classes extending `Exception` so I know that the message is always there - with simplified code 
```
class CustomExceptionMixin:
    message = ""

    def __init__(
        self, message: str, *args, **kwargs
    ) -> None:
        self.message = message
        super().__init__(message, **kwargs)  # type: ignore
```

**Comment by BioBox** [comment]
I'm also still having this error, but in my case it's from mixins with `pathlib`s classes.

**Comment by ziyuang** [comment]
> My usecase is adding mixins to classes extending `Exception` so I know that the message is always there - with simplified code

Have the same use case and the issue still lingers with `mypy==1.9.0`.
